### PR TITLE
Change parametrization of AutoLowRankMultivariateNormal

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2120 

This aims to improve convergence of `AutoLowRankMultivariateNormal` by separating "noise scale" parameters from "noise direction" parameters. Intuitively this allows the uncertainty to grow slowly early in learning, when the autoguide is still learning the loc parameter; thus this PR is a smooth way to accomplish the staged learning of #2113 .

The old parametrization was:
```py
cov = diagonal + cov_factor @ cov_factor.T
```
The new parametrization is:
```py
cov = scale.unsqueeze(-1) * (eye + cov_factor @ cov_factor.T) * scale
    = scale.pow(2).diag_embed() + scaled_cov_factor @ scaled_cov_factor
```
where
```py
scaled_cov_factor = scale.unsqueeze(-1) * cov_factor
```
### Experiment 1

I have tested this on a high-dimensional variational inference problem (dim=22407, rank=100, real world data) and found this PR to stabilize training:
<details>

old loss curve (which required custom per-parameter learning rates):
image
new loss curve (with higher learning rate):
image

</details>

## Experiment 2

This example demonstrates more stable convergence on a synthetic dataset:
https://github.com/pyro-ppl/sandbox/blob/auto-lowrank-mvn/2019-11-lowrank/experiment.ipynb

<details>

image

image

</details>

## Experiment 3

This example demonstrates identification of true low rank structure when the model matches the guide. https://github.com/pyro-ppl/sandbox/blob/auto-lowrank-mvn/2019-11-lowrank/exact.ipynb

<details>

image

</details>

## Experiment 4

This is like experiment 3, but when dimensions are randomly scaled by a scale in [exp(-1), exp(1)]. https://github.com/pyro-ppl/sandbox/blob/auto-lowrank-mvn/2019-11-lowrank/exact-scaled.ipynb

<details>

image

</details>

## Tested
- tested on experiments
- existing behavior is covered by unit tests

### CODE REVIEW

File: pyro/infer/autoguide/guides.py [hunk 1]
```diff
@@ -588,20 +588,22 @@ def get_posterior(self, *args, **kwargs):
...
 
     def _loc_scale(self, *args, **kwargs):
         loc = pyro.param("{}_loc".format(self.prefix))
+        scale = pyro.param("{}_scale".format(self.prefix))
         factor = pyro.param("{}_cov_factor".format(self.prefix))
-        diagonal = pyro.param("{}_cov_diag".format(self.prefix))
-        scale = (factor.pow(2).sum(-1) + diagonal).sqrt()
-        return loc, scale
+        scale = scale * (factor.pow(2).sum(-1) + 1).sqrt()
+        return loc, scale * 2
```
Comment (martinjankowiak) [inline pyro/infer/autoguide/guides.py]: why 2?

Comment (fritzo) [reply pyro/infer/autoguide/guides.py]: oops, fixed. The 2 was a mutation test: I added temporarily just to ensure tests were strong enough to catch that error. Then I forgot to revert the mutation.

**Comment by fehiepsi** [comment]
Bring `diagonal` to `scale` make sense to me, but I am not persuaded that using scaled_cov_factor is better. Just to confirm (though I thought that you have tested it), could you plot the learning curve if we use `scale` and `cov_factor`?

**Comment by fritzo** [comment]
@fehiepsi I'll add some more comprehensive studies tomorrow.

**Comment by fritzo** [comment]
@fehiepsi I have added another experimental study, see notebook.

>  could you plot the learning curve if we use scale and cov_factor?

Can you explain? If we change only `cov_diagonal -> scale`, and leave `cov_factor` as-is, results should be identical to existing Pyro dev (proof: `cov_diagonal` is constrained by constraints.positive, which transformed via `ExpTransform`. Hence the `cov_diagonal` and `scale` parameters will differ by a shift of `log(2)`. Since they only differ by a shift, their gradients will be identical. After identical gradient updates, they will still differ by a shift; hence loss will be identical. []).

**Comment by fritzo** [comment]
I'd like to add more tests to ensure this fits the true covariance of a true man...

**Comment by fritzo** [comment]
OK, I have added another experiment. I am pretty convinced this parametrization is superior. Can reviewers think of any other experiments we might try to provide further evidence?

**Comment by martinjankowiak** [comment]
experiments certainly convince me!

File: pyro/infer/autoguide/guides.py [hunk 2]
```diff
@@ -588,19 +588,21 @@ def get_posterior(self, *args, **kwargs):
         Returns a LowRankMultivariateNormal posterior distribution.
         """
         loc = pyro.param("{}_loc".format(self.prefix), self._init_loc)
+        scale = pyro.param("{}_scale".format(self.prefix),
+                           lambda: loc.new_full((self.latent_dim,), self._init_scale),
```
Comment (fehiepsi) [inline pyro/infer/autoguide/guides.py]: do you want to make the init values `0.5 ** 0.5 * self._init_scale` so that initial `self._loc_scale()[1] == self._init_scale`?

File: pyro/infer/autoguide/guides.py [hunk 3]
```diff
@@ -588,19 +588,21 @@ def get_posterior(self, *args, **kwargs):
...
+                           constraint=constraints.positive)
         factor = pyro.param("{}_cov_factor".format(self.prefix),
                             lambda: loc.new_empty(self.latent_dim, self.rank).normal_(
-                                0, self._init_scale * (0.5 / self.rank) ** 0.5))
-        diagonal = pyro.param("{}_cov_diag".format(self.prefix),
-                              lambda: loc.new_full((self.latent_dim,), 0.5 * self._init_scale ** 2),
-                              constraint=constraints.positive)
+                                0, self._init_scale / self.rank ** 0.5))
```
Comment (fehiepsi) [inline pyro/infer/autoguide/guides.py]: I think you want `.normal_(0, 1 / self.rank ** 0.5)` here.

File: pyro/infer/autoguide/guides.py (hunk 2 above)
Comment (fritzo) [reply pyro/infer/autoguide/guides.py]: done.

File: pyro/infer/autoguide/guides.py (hunk 3 above)
Comment (fritzo) [reply pyro/infer/autoguide/guides.py]: done.

**Comment by fritzo** [comment]
Let's hold off merging this until after #2126, so as to avoid merge conflicts.

@fehiepsi thanks for the careful review!

**Comment by fritzo** [comment]
I believe this is ready to merge.
